pkg/shares/share_merging.go

```diff
@@ -176,5 +176,58 @@ func ParseShares(rawShares [][]byte) ([]ShareSequence, error) {
 		sequences = append(sequences, currentSequence)
 	}
 
+	for _, sequence := range sequences {
+		if err := sequence.validSequenceLength(); err != nil {
+			return sequences, err
+		}
+	}
+
 	return sequences, nil
 }
+
+// validSequenceLength extracts the sequenceLength written to the first share
+// and returns an error if the number of shares needed to store a sequence of
+// length sequenceLength doesn't match the number of shares in this share
+// sequence. Returns nil if there is no error.
+func (s ShareSequence) validSequenceLength() error {
+	if len(s.Shares) == 0 {
+		return nil
+	}
```

Comment:
[question]
is any sequence of length 0 valid? I was always under the impression that it was explicitly invalid

Reply:
Good question, I think you're right that it should always be explicitly invalid. I'll add a unit test for this edge case.

Reply:
Resolved in https://github.com/celestiaorg/celestia-app/pull/865/commits/02905fc521eda72bcb152ce95a28384270c783ee

Note, a unit test wasn't added for this case b/c its impossible to reach this code path without finagling with internals of `ParseShares` because `validSequenceLength` won't be called if no sequences exist